test/gtest/core/test_mc_reduce.cc

```diff
@@ -0,0 +1,257 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2021.  ALL RIGHTS RESERVED.
+ * See file LICENSE for terms.
+ */
+
+extern "C" {
+#include <core/ucc_mc.h>
+}
+#include <common/test.h>
+
+template<ucc_datatype_t, template <typename P> class op>
+struct ReductionTest;
+
+template<template <typename P> class op>
+struct ReductionTest<UCC_DT_INT16, op>
+{
+    using type = int16_t;
+    const static ucc_datatype_t dt = UCC_DT_INT16;
+    const static ucc_reduction_op_t redop = op<type>::redop;
+    static void assert_equal(type arg1, type arg2) {
+        ASSERT_EQ(arg1, arg2);
+    }
+    static type do_op(type arg1, type arg2) {
+        op<type> _op;
+        return _op(arg1, arg2);
+    }
+};
+
+template<template <typename P> class op>
+struct ReductionTest<UCC_DT_INT32, op>
+{
+    using type = int32_t;
+    const static ucc_datatype_t dt = UCC_DT_INT32;
+    const static ucc_reduction_op_t redop = op<type>::redop;
+    static void assert_equal(type arg1, type arg2) {
+        ASSERT_EQ(arg1, arg2);
+    }
+    static type do_op(type arg1, type arg2) {
+        op<type> _op;
+        return _op(arg1, arg2);
+    }
+};
+
+template<template <typename P> class op>
+struct ReductionTest<UCC_DT_INT64, op>
```

Comment:
Maybe make a macro to declare this template class: the code is identical almost entirely:
#define REDUCTION_TEST_DECLARE(_type, _ucc_type, _assert_eq) do {
using type = _type; \
const status ucc_datatype_t dt = _ucc_type; \
ucc_reduction_op redop = op<type>::redop;\
static void assert_equal(...) { \
_assert_eq(arg1, arg2); \
} \
...
} while (0)

and then 
REDUCTION_TEST_DECLARE(int64, UCC_DT_INT64, ASSERT_EQ);
REDUCTION_TEST_DECLARE(float, UCC_DT_FLOAT32, ASSERT_EQ_FLOAT);

etc

Reply:
Sure, it's just not easy to debug when test is wrapped into macro. But I want first extend test with memory type and add reduce multi.

Reply:
we can keep it as is if you think it makes inconvenient debug

Reply:
i mean debug of test itself, once all tests parameters are there (reduce, reduce multi, mem types) we can wrap class declaration in macro